doko/match/extrapoints.go

```diff
@@ -18,39 +18,74 @@ func findExtraPoints(g *game.Game) (ans []ExtraPoint) {
 	if !extraPointsApply(g.Mode.Type()) {
 		return
 	}
-	ans = append(ans, doppelkoepfe(g)...)
-	ans = append(ans, fuechse(g)...)
-	ans = append(ans, karlchen(g)...)
+	ans = append(ans, doppelkopfs(g)...)
+	ans = append(ans, foxesCaught(g)...)
+	ans = append(ans, charlie(g)...)
 	return
 }
 
-func doppelkoepfe(game *game.Game) []ExtraPoint {
+func EventsOccured(g *game.Game, trick *game.Trick) (ans []ExtraPointType) {
+	if !extraPointsApply(g.Mode.Type()) {
+		return
+	}
+	if scoresDoppelkopf(trick) {
+		ans = append(ans, Doppelkopf)
+	}
+	for _, player := range game.Players() {
+		if maybeFoxCaught(player, trick) {
+			ans = append(ans, FoxCaught)
+		}
+	}
+	if charlieWinLastTrick(g) {
+		ans = append(ans, Charlie)
+	}
+	return ans
+}
+
+func scoresDoppelkopf(trick *game.Trick) bool {
+	return trick.Score() >= 40
+}
+
+func doppelkopfs(game *game.Game) []ExtraPoint {
 	var ans []ExtraPoint
 	for i, trick := range game.CompleteTricks {
-		if trick.Score() >= 40 {
+		if scoresDoppelkopf(&trick) {
 			ep := ExtraPoint{Doppelkopf, trick.Winner, i}
 			ans = append(ans, ep)
 		}
 	}
 	return ans
 }
 
-func fuechse(g *game.Game) []ExtraPoint {
+//logic without party information
+func maybeFoxCaught(player game.Player, trick *game.Trick) bool {
```

Comment:
note: in the [doppelkopf app I use regularly](http://isar-interactive.de/doppelkopf.html), party information is consirede public once a player has played a queen of clubs, which is then used to emit the „fox caught“ event more precisely. However in real games it is not uncommon to forget who played a »Dulle«, I think this should also be possibility in the app. ;-)
But at least when contra/re etc. are announced that information could be used (in a separate PR, of course) to add constraints to this function.